nova_vm/src/ecmascript/builtins/control_abstraction_objects/iteration/iterator_prototype.rs

```diff
@@ -34,15 +50,90 @@ impl IteratorPrototype {
         Ok(this_value.unbind())
     }
 
+    /// ### [27.1.4.7 Iterator.prototype.forEach ( procedure )](https://tc39.es/ecma262/#sec-iterator.prototype.foreach)
+    fn for_each<'gc>(
+        agent: &mut Agent,
+        this_value: Value,
+        arguments: ArgumentsList,
+        mut gc: GcScope<'gc, '_>,
+    ) -> JsResult<Value<'gc>> {
+        let nogc = gc.nogc();
+        let this_value = this_value.bind(nogc);
+        let procedure = arguments.get(0).bind(nogc);
+
+        // 1. Let O be the this value.
+        // 2. If O is not an Object, throw a TypeError exception.
+        let Ok(o) = Object::try_from(this_value) else {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "'this' is not an object",
+                nogc,
+            ));
+        };
+
+        // 3. Let iterated be the Iterator Record { [[Iterator]]: O, [[NextMethod]]: undefined, [[Done]]: false }.
+        // 4. If IsCallable(procedure) is false, then
+        let Some(procedure) = is_callable(procedure, nogc).unbind().bind(gc.nogc()) else {
+            // a. Let error be ThrowCompletion(a newly created TypeError object).
+            let error = agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "'procedure' is not callable",
+                nogc,
+            );
+            // b. Return ? IteratorClose(iterated, error).
+            return iterator_close(agent, o.unbind(), Err(error), gc);
+        };
+        let scoped_procedure = Value::from(procedure).scope(agent, gc.nogc());
+
+        // 5. Set iterated to ? GetIteratorDirect(O).
+        // if i rebind `iterated` here borrow checker complains
+        let iterated = get_iterator_direct(agent, o.unbind(), gc.reborrow()).unbind()?;
+        let Some(iterated) = iterated else {
+            // what should happen here?
+            todo!();
```

Comment:
The spec isn't clear about what should happen here. Should it just be a `TypeError` like the above for `procedure`?